### PR TITLE
Follow up to dependency verification using signatures

### PR DESCRIPTION
**Pull request by melix**
### Context

Some improvements on signature verification, in particular trusted keys, built on top of #11712

### CODE REVIEW

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationSignatureCheckIntegTest.groovy [hunk 1]
```diff
@@ -528,4 +528,30 @@ This can indicate that a dependency has been compromised. Please verify carefull
...
+                implementation "org:foo:1.0"
+            }
+        """
+
+        when:
+        fails ":compileJava"
+
+        then:
+        failure.assertHasCause """Dependency verification failed for configuration ':compileClasspath':
+  - On artifact foo-1.0.jar (org:foo:1.0): expected a 'sha256' checksum of 'nope' but was '20ae575ede776e5e06ee6b168652d11ee23069e92de110fdec13fbeaa5cf3bbc'
```
Comment (ljacomet) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationSignatureCheckIntegTest.groovy]: How complicated would it be to clearly state we attempted a signature check but could not.
Which is different than having a module configured to ignore signature, or not having signature check enabled at all.
Makes sense?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/ChecksumAndSignatureVerificationOverride.java [hunk 2]
```diff
@@ -97,12 +104,44 @@ private static URI uri(String uri) {
...
-            synchronized (failures) {
-                failures.put(artifact, f);
+    public void onArtifact(ArtifactKind kind, ModuleComponentArtifactIdentifier artifact, File mainFile, Factory<File> signatureFile) {
+        VerificationEvent event = new VerificationEvent(kind, artifact, mainFile, signatureFile);
+        synchronized (verificationEvents) {
+            verificationEvents.add(event);
+        }
+    }
+
+    private void verifyConcurrently() {
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/ChecksumAndSignatureVerificationOverride.java]: From what I see, this delayed verification still works in a way that the artifacts cannot be accessed outside from the logic downloading them.
But I wonder if we can guarantee that.
And delaying verification to after something has been used would defeat the purpose IMO.

Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/ChecksumAndSignatureVerificationOverride.java]: This hasn't changed. We perform verification _as soon as an artifact is accessed_. The only thing that changed is that _all artifacts_ are checked concurrently.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationSignatureCheckIntegTest.groovy (hunk 1 above)
Comment (melix) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationSignatureCheckIntegTest.groovy]: Yes, I'll try to improve this.

**Comment by melix** [comment]
@ljacomet PTAL

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationSignatureCheckIntegTest.groovy (hunk 1 above)
Comment (melix) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationSignatureCheckIntegTest.groovy]: Done

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationIntegrityCheckIntegTest.groovy [hunk 3]
```diff
@@ -108,9 +108,9 @@ class DependencyVerificationIntegrityCheckIntegTest extends AbstractDependencyVe
 
         then:
         failure.assertHasCause("""Dependency verification failed for configuration ':compileClasspath':
-  - On artifact foo-1.0.jar (org:foo:1.0): expected a '$kind' checksum of 'invalid' but was '$value'
-  - Artifact foo-1.0.pom (org:foo:1.0) checksum is missing from verification metadata.
-This can indicate that a dependency has been compromised. Please verify carefully the checksums.""")
+  - On artifact foo-1.0.jar (org:foo:1.0) in repository 'maven': expected a '$kind' checksum of 'invalid' but was '$value'
+  - On artifact foo-1.0.pom (org:foo:1.0) in repository 'maven': checksum is missing from verification metadata.
+This can indicate that a dependency has been compromised. Please carefully verify the checksums.""")
```
Comment (ljacomet) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationIntegrityCheckIntegTest.groovy]: Really nice additions to the error messages there! 👍

**Pull request merged**